libslide/src/bignum/add.rs

```diff
@@ -0,0 +1,135 @@
+#![allow(clippy::suspicious_arithmetic_impl)]
+use crate::bignum::Bignum;
+use std::mem;
+use std::ops;
+
+impl ops::Add for Bignum {
+    type Output = Bignum;
+
+    fn add(self, rhs: Bignum) -> Bignum {
+        let is_neg: bool = rhs.is_neg && self.is_neg;
+
+        // this assumes the signs are both positive for now
+        let mut carry: u8 = 0;
+        let mut lhs_size: usize = self.dec.len();
+        let mut rhs_size: usize = rhs.dec.len();
+        let mut res_dec: Vec<u8> = Vec::new();
+
+        // 1. Handle Decimal
+
+        // swap if neccessary to make lhs smaller. Initialize big vec as the bigger vector
+        let big_vec = if lhs_size > rhs_size {
+            mem::swap(&mut lhs_size, &mut rhs_size);
+            self.dec.clone()
+        } else {
+            rhs.dec.clone()
+        };
+
+        // compute the decimal addition up to the size of the small vec
+        let mut res: u8;
+        for i in (0..lhs_size).rev() {
+            res = self.dec[i] + rhs.dec[i] + carry;
+            res_dec.push(res % 10);
+            carry = (res > 9) as u8;
+        }
+
+        // decimals are added to the vec in reverse order so we switch order back
+        res_dec = res_dec.iter().rev().cloned().collect();
+
+        // add the remaining decimals of the larger vector
+        let mut i = lhs_size;
+        while i < rhs_size {
+            res_dec.push((big_vec[i]) % 10);
+            i += 1;
+        }
+
+        // 2. Handle Integer
+
+        lhs_size = self.int.len();
+        rhs_size = rhs.int.len();
+
+        // we reverse the vectors since our carry matters when we handle
+        // adding the larger vector
```

Comment:
Maybe we should store the integer part in the BigNum with the LSB at the 0 index. Otherwise we are reversing twice each time we do any operation, which is way more expensive than reversing once for formatting (which is expensive anyway)

Reply:
these comments are wrong and should've been deleted on my latest refactor. I rewrote the algo so it doesn't need to reverse the initial vectors. It should be the min time complexity of O(lhs+rhs)

Reply:
Still, we reverse the vector at the end.

Reply:
even if it was built the opposite way, then we would have to reverse the dec vector at the end

Reply:
Why?

Reply:
You wouldn't have to reverse it at all

Reply:
If you keep 123.456 as
int: 321
dec: 456

Reply:
I see.Let me finish subtract and ill refactor.